### PR TITLE
drivers: dma: stm32 dmamux device is passed to the dma device

### LINKED ISSUE #29992: dma tests fail with stm32wb55 and stm32l476  nucleo boards
**Describe the bug**
running  `tests/drivers/dma/chan_blen_transfer` and `tests/drivers/dma/loop_transfer` fail
on nucleo_wb55rg when boards/nucleo_wb55rg.conf and boards/nucleo_wb55rg.overlay are added

running  `tests/drivers/dma/chan_blen_transfer` fails
on nucleo_l476rg 

(maybe others)

**To Reproduce**
.$ west build -p auto -b nucleo_wb55rg tests/drivers/dma/chan_blen_transfer/
 See error
```
Running test suite dma_m2m_test
===================================================================
START - test_dma_m2m_chan0_burst8
Cannot get dma controller

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/dma/chan_blen_transfere

 FAIL - test_dma_m2m_chan0_burst8
===================================================================
START - test_dma_m2m_chan1_burst8
Cannot get dma controller

    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/dma/chan_blen_transfere
                                                                                
 FAIL - test_dma_m2m_chan1_burst8                                               
===================================================================             
START - test_dma_m2m_chan0_burst16                                              
Cannot get dma controller                                                       
                                                                                
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/dma/chan_blen_transfere
                                                                                
 FAIL - test_dma_m2m_chan0_burst16                                              
===================================================================             
START - test_dma_m2m_chan1_burst16                                              
Cannot get dma controller                                                       
                                                                                
    Assertion failed at WEST_TOPDIR/zephyr/tests/drivers/dma/chan_blen_transfere
                                                                                
 FAIL - test_dma_m2m_chan1_burst16                                              
===================================================================             
Test suite dma_m2m_test failed.                                                 
===================================================================             
PROJECT EXECUTION FAILED     

```

.$ west build -p auto -b nucleo_wb55rg tests/drivers/dma/loop_transfer/
 See error
```
DMA memory to memory transfer started on DMAMUX_1                               
Preparing DMA Controller                                                        
Starting the transfer and waiting for 1 second                                  
TX data: The quick brown fox jumps over the lazy dog ....                       
block_size 48                                                                   
E: cannot configure the dma stream -1.                                          
ERROR: unfinished transfer                                                      
Each RX buffer should contain the full TX buffer string.                        
TX data: The quick brown fox jumps over the lazy dog ....                       
RX data Loop 0: The quick brown fox jumps over the lazy dog ....                
RX data Loop 1: The quick brown fox jumps over the lazy dog ....                
RX data Loop 2:                                                                 
RX data Loop 3:                                                                 
RX data Loop 4:                                                                 
Finished: DMA       
```                                                            

**Expected behavior**
tests PASSED

**Additional context**
master branch : a7aef46758597af36cd64a08d07711fd120d52a1

### PR DESCRIPTION
**Pull request by FRASTM**
The dmamux device reference is passed to the associated dma device during the dmamux_stm32_configure() 
through the user_data field of the dma_config structure.
So that, later on the incoming dma irq, the dma knows the dmamux where to route its callback function

This is demonstrated using the `tests/drivers/dma/loop_transfer` and `tests/drivers/dma/chan_blen_transfer` running on the stm32wb55 nucleo board.

fixes https://github.com/zephyrproject-rtos/zephyr/issues/29992

### CODE REVIEW

**Comment by FRASTM** [comment]
testing  tests/drivers/dma/loop_transfer on nucleo_wb55rg:
```
*** Booting Zephyr OS build zephyr-v2.4.0-1463-g44c480e8d157  ***                                                    
DMA memory to memory transfer started on DMAMUX_1                                                                    
Preparing DMA Controller                                                                                             
Starting the transfer and waiting for 1 second                                                                       
TX data: The quick brown fox jumps over the lazy dog ....                                                            
block_size 48                                                                                                        
Each RX buffer should contain the full TX buffer string.                                                             
TX data: The quick brown fox jumps over the lazy dog ....                                                            
RX data Loop 0: The quick brown fox jumps over the lazy dog ....                                                     
RX data Loop 1: The quick brown fox jumps over the lazy dog ....                                                     
RX data Loop 2: The quick brown fox jumps over the lazy dog ....                                                     
RX data Loop 3: The quick brown fox jumps over the lazy dog ....                                                     
RX data Loop 4: The quick brown fox jumps over the lazy dog ....                                                     
Finished: DMA 
```

File: tests/drivers/dma/loop_transfer/boards/nucleo_wb55rg.conf [hunk 1]
```diff
@@ -0,0 +1,4 @@
+CONFIG_DMA_LOOP_TRANSFER_DRV_NAME="DMAMUX_1"
+CONFIG_DMA_LOOP_TRANSFER_CHANNEL_NR=11
+CONFIG_DMA_MCUX_TEST_SLOT_START=0
```
Comment (erwango) [inline tests/drivers/dma/loop_transfer/boards/nucleo_wb55rg.conf:3]: This is a MCUX ("depends on DMA_MCUX_EDMA and depends on (SOC_SERIES_KINETIS_K6X || SOC_SERIES_KINETIS_KE1XF)") specific property.  We should not be able to use it for STM32.

If requested to pass the test, then the test should be reworked.

File: tests/drivers/dma/loop_transfer/boards/nucleo_wb55rg.overlay [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2020 Erwin Rol <erwin@erwinrol.com>
```
Comment (erwango) [inline tests/drivers/dma/loop_transfer/boards/nucleo_wb55rg.overlay:2]: Use your own copyright

File: tests/drivers/dma/loop_transfer/boards/nucleo_l476rg.overlay [hunk 3]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2020 Erwin Rol <erwin@erwinrol.com>
```
Comment (erwango) [inline tests/drivers/dma/loop_transfer/boards/nucleo_l476rg.overlay:2]: Use your own copyright

File: tests/drivers/dma/loop_transfer/boards/nucleo_l476rg.conf [hunk 4]
```diff
@@ -0,0 +1,3 @@
+CONFIG_DMA_LOOP_TRANSFER_DRV_NAME="DMA_1"
+CONFIG_DMA_LOOP_TRANSFER_CHANNEL_NR=4
+CONFIG_DMA_MCUX_TEST_SLOT_START=0
```
Comment (erwango) [inline tests/drivers/dma/loop_transfer/boards/nucleo_l476rg.conf:3]: Cf previous comment

File: tests/drivers/dma/chan_blen_transfer/Kconfig [hunk 5]
```diff
@@ -0,0 +1,18 @@
+# Copyright (c) 2020 Erwin Rol <erwin@erwinrol.com>
```
Comment (erwango) [inline tests/drivers/dma/chan_blen_transfer/Kconfig:1]: Ditto

File: tests/drivers/dma/chan_blen_transfer/boards/nucleo_l476rg.conf [hunk 6]
```diff
@@ -0,0 +1,4 @@
+CONFIG_DMA_TRANSFER_DRV_NAME="DMA_1"
+CONFIG_DMA_TRANSFER_CHANNEL_NR_0=4
+CONFIG_DMA_TRANSFER_CHANNEL_NR_1=2
+CONFIG_DMA_MCUX_TEST_SLOT_START=0
```
Comment (erwango) [inline tests/drivers/dma/chan_blen_transfer/boards/nucleo_l476rg.conf:4]: ditto

File: tests/drivers/dma/chan_blen_transfer/boards/nucleo_l476rg.overlay [hunk 7]
```diff
@@ -0,0 +1,13 @@
+/*
+ * Copyright (c) 2020 Erwin Rol <erwin@erwinrol.com>
```
Comment (erwango) [inline tests/drivers/dma/chan_blen_transfer/boards/nucleo_l476rg.overlay:2]: ditto

File: tests/drivers/dma/chan_blen_transfer/boards/nucleo_wb55rg.conf [hunk 8]
```diff
@@ -0,0 +1,3 @@
+CONFIG_DMA_LOOP_TRANSFER_DRV_NAME="DMAMUX_1"
+CONFIG_DMA_LOOP_TRANSFER_CHANNEL_NR=11
+CONFIG_DMA_MCUX_TEST_SLOT_START=0
```
Comment (erwango) [inline tests/drivers/dma/chan_blen_transfer/boards/nucleo_wb55rg.conf:3]: ditto

File: tests/drivers/dma/chan_blen_transfer/boards/nucleo_wb55rg.overlay [hunk 9]
```diff
@@ -0,0 +1,17 @@
+/*
+ * Copyright (c) 2020 Erwin Rol <erwin@erwinrol.com>
```
Comment (erwango) [inline tests/drivers/dma/chan_blen_transfer/boards/nucleo_wb55rg.overlay:2]: ditto

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
Please clear CONFIG_DMA_MCUX_TEST_SLOT_START from the commit messages.
One copyright remains to be fixed.

Otherwise LGTM

File: tests/drivers/dma/loop_transfer/boards/nucleo_wb55rg.conf (hunk 1 above)
Comment (FRASTM) [reply tests/drivers/dma/loop_transfer/boards/nucleo_wb55rg.conf]: removed

File: tests/drivers/dma/loop_transfer/boards/nucleo_l476rg.overlay (hunk 3 above)
Comment (FRASTM) [reply tests/drivers/dma/loop_transfer/boards/nucleo_l476rg.overlay]: done : `Copyright (c) 2020 STMicroelectronics`

File: tests/drivers/dma/chan_blen_transfer/Kconfig (hunk 5 above)
Comment (erwango) [reply tests/drivers/dma/chan_blen_transfer/Kconfig:1]: This one remains to be fixed

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — APPROVED** [review]

**Review by ABOSTM — APPROVED** [review]
LGTM

**Comment by FRASTM** [comment]
this PR must be reworked as the _tests/drivers/spi/spi_loopback_ now fails:
```
START - test_spi_loopback                                                       
I: SPI test on buffers TX/RX 0x2000019d/0x20000e9a                              
I: SPI test slow config                                                         
I: Start complete multiple                                                      
I: CS control inhibited (no GPIO device)                                        
E: DMA callback channel 11 is not valid.                                        
E: ***** MPU FAULT *****                                                        
E:   Data Access Violation   
```

**Comment by erwango** [comment]
@FRASTM please let us know when everything is fine

**Comment by FRASTM** [comment]
With this fix, only the tests/drivers/dma applications are modified to include the support of the DMAMUX on some stm32 devices (typ. stm32wb55rg). In this case, the callback function (`dma_user_callback`) of the tests/drivers/dma/loop_back must refer to the DMAMUX (not DMA) device, even if the irq is trigger by the DMA. This DMAMUX device is stored in the `dma_cfg.user_data` during the application init. At this time, the dma devices refers to the DMAMUX instance.

**Comment by FRASTM** [comment]
with this fix and connecting SPI 1 MISO to MOSI  (PA7 = PA6 on CN10 of the nucleo_wb55rg board)

```
*** Booting Zephyr OS build zephyr-v2.4.0-1767-gf46d3a888154  ***               
Running test suite test_spi                                                     
===================================================================             
START - test_spi_loopback                                                       
I: SPI test on buffers TX/RX 0x200001a1/0x20000ea6                              
I: SPI test slow config                                                         
I: Start complete multiple                                                      
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: Start complete loop                                                          
I: Passed                                                                       
I: Start null tx                                                                
I: Passed                                                                       
I: Start half start                                                             
I: Passed                                                                       
I: Skip half end                                                                
I: Skip every 4                                                                 
I: SPI test fast config                                                         
I: Start complete multiple                                                      
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: Start complete loop                                                          
I: Passed                                                                       
I: Start null tx                                                                
I: Passed                                                                       
I: Start half start                                                             
I: Passed                                                                       
I: Skip half end                                                                
I: Skip every 4                                                                 
I: Start complete loop                                                          
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: Start complete loop                                                          
I: CS control inhibited (no GPIO device)                                        
I: Passed                                                                       
I: All tx/rx passed                                                             
 PASS - test_spi_loopback                                                       
===================================================================             
Test suite test_spi succeeded                                                   
===================================================================             
PROJECT EXECUTION SUCCESSFUL  
```

**Review by erwango — CHANGES_REQUESTED** [review]
LGTM, but please add "dma" support in board yaml files, so this can be tested in CI.

**Review by erwango — APPROVED** [review]

**Pull request merged**